lib/upload_to_s3.js

```diff
@@ -1,45 +1,87 @@
 import https from 'https';
 import url from 'url';
+import fs from 'fs';
+import temporary from 'temporary';
+import promiseRetry from 'promise-retry';
+import _ from 'lodash';
 
-export default async function uploadToS3 (task, sourceStream, artifactName, expiration, httpsHeaders) {
-  var queue = task.runtime.queue;
-
-  var artifact = await queue.createArtifact(
-    task.status.taskId,
-    task.runId,
-    artifactName,
-    {
-      // Why s3? It's currently cheaper to store data in s3 this could easily
-      // be used with azure simply by changing s3 -> azure.
-      storageType: 's3',
-      expires: new Date(Math.min(expiration, new Date(task.task.expires))),
-      contentType: httpsHeaders['content-type']
+export default async function uploadToS3 (
+  task,
+  sourceStream,
+  artifactName,
+  expiration,
+  httpsHeaders,
+  putUrl,
+  httpOptions)
+{
+  let queue = task.runtime.queue;
+
+  let tmp = new temporary.File();
+
+  try {
+    await new Promise((accept, reject) => {
+      let stream = fs.createWriteStream(tmp.path);
+      stream.on('error', reject);
+      stream.on('finish', accept);
+      sourceStream.pipe(stream);
+    });
+
+    if (!putUrl) {
+      var artifact = await queue.createArtifact(
+        task.status.taskId,
+        task.runId,
+        artifactName,
+        {
+          // Why s3? It's currently cheaper to store data in s3 this could easily
+          // be used with azure simply by changing s3 -> azure.
+          storageType: 's3',
+          expires: new Date(Math.min(expiration, new Date(task.task.expires))),
+          contentType: httpsHeaders['content-type']
+        }
+      );
+
+      putUrl = artifact.putUrl;
     }
-  );
-
-  let parsedUrl = url.parse(artifact.putUrl);
-  let options = {
-    hostname: parsedUrl.hostname,
-    path: parsedUrl.path,
-    method: 'PUT',
-    headers: httpsHeaders
-  };
-
-  let req = https.request(options);
-  sourceStream.pipe(req);
-
-  let response = await new Promise((accept, reject) => {
-    req.on('response', accept);
-    req.on('error', reject);
-    req.setTimeout(5 * 60 * 1000, reject);
-  });
-
-  // Flush the data from the reponse so it's not held in memory
-  response.resume();
-
-  if (response.statusCode !== 200) {
-    throw new Error(
-      `Could not upload artifact. ${response.error} Status Code: ${response.statusCode}`
-    );
+
+    var parsedUrl = url.parse(putUrl);
+    var options = _.defaults({
+      hostname: parsedUrl.hostname,
+      path: parsedUrl.path,
+      method: 'PUT',
+      headers: httpsHeaders,
+      port: parsedUrl.port
+    }, httpOptions);
+
+    // promiseRetry defaults to 10 attempts before failing
+    await promiseRetry((retry, number) => {
+      if (number > 1) { // if it's not the first attempt
+        task.runtime.log('retrying artifact upload');
```

Comment:
Could we add some information here about the task ID, run ID, and artifact name that's being retried?